File: hello-csharp/src/Server/MmapCache.cs
using System.IO.MemoryMappedFiles;

namespace AudioStreamServer;

/// <summary>File-backed memory-mapped cache for a single stream.</summary>
public sealed class MmapCache : IDisposable
{
    private const long DefaultPageSize = 64L * 1024 * 1024;   // 64 MB
    private const long MaxCacheSize = 8L * 1024 * 1024 * 1024; // 8 GB

    private readonly string _path;
    private MemoryMappedFile? _mmf;
    private MemoryMappedViewAccessor? _accessor;
    private long _capacity;
    private bool _disposed;

    public MmapCache(string path)
    {
        _path = path;
    }

    /// <summary>Create the backing file and open the memory map.</summary>
    public void Create(long initialSize = 0)
    {
        var size = Math.Max(initialSize, DefaultPageSize);
        using (var fs = new FileStream(_path, FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite))
        {
            fs.SetLength(size);
        }
        Open(size);
    }

    /// <summary>Write data at the given offset, growing the file if needed.</summary>
    public void Write(long offset, byte[] data)
    {
        var required = offset + data.Length;
        if (required > _capacity) Resize(required);
        _accessor!.WriteArray(offset, data, 0, data.Length);
    }

    /// <summary>Read data from the cache.</summary>
    public byte[] Read(long offset, int length)
    {
        EnsureOpen();
        var buf = new byte[length];
        _accessor!.ReadArray(offset, buf, 0, length);
        return buf;
    }

    /// <summary>Truncate the file to its final size and flush.</summary>
    public void Finalize(long finalSize)
    {
        Close();
        using (var fs = new FileStream(_path, FileMode.Open, FileAccess.ReadWrite))
        {
            fs.SetLength(finalSize);
        }
        Open(finalSize);
        _accessor!.Flush();
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        Close();
    }

    // ── Internal ────────────────────────────────────────────────────────

    private void Open(long size)
    {
        _mmf = MemoryMappedFile.CreateFromFile(_path, FileMode.Open, null, size, MemoryMappedFileAccess.ReadWrite);
        _accessor = _mmf.CreateViewAccessor(0, size, MemoryMappedFileAccess.ReadWrite);
        _capacity = size;
    }

    private void EnsureOpen()
    {
        if (_mmf != null) return;
        if (!File.Exists(_path)) throw new FileNotFoundException("Cache file not found", _path);
        var size = new FileInfo(_path).Length;
        Open(size);
    }

    private void Close()
    {
        _accessor?.Dispose();
        _accessor = null;
        _mmf?.Dispose();
        _mmf = null;
    }

    private void Resize(long required)
    {
        var newSize = _capacity;
        while (newSize < required) newSize = Math.Min(newSize * 2, MaxCacheSize);
        if (newSize > MaxCacheSize) throw new InvalidOperationException("Cache exceeds max size");
        Close();
        using (var fs = new FileStream(_path, FileMode.Open, FileAccess.ReadWrite))
        {
            fs.SetLength(newSize);
        }
        Open(newSize);
    }
}

File: hello-kotlin/src/main/kotlin/server/MmapCache.kt
package server

import java.io.RandomAccessFile
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.Paths

/** File-backed memory-mapped cache for a single stream. */
class MmapCache(private val path: String) : AutoCloseable {

    companion object {
        private const val DEFAULT_PAGE_SIZE = 64L * 1024 * 1024  // 64 MB
        private const val MAX_CACHE_SIZE = 8L * 1024 * 1024 * 1024 // 8 GB
    }

    private var file: RandomAccessFile? = null
    private var channel: FileChannel? = null
    private var buffer: MappedByteBuffer? = null
    private var capacity: Long = 0

    fun create(initialSize: Long = 0) {
        val size = maxOf(initialSize, DEFAULT_PAGE_SIZE)
        val filePath = Paths.get(path)
        if (Files.exists(filePath)) Files.delete(filePath)
        RandomAccessFile(path, "rw").use { it.setLength(size) }
        open(size)
    }

    fun write(offset: Long, data: ByteArray) {
        val required = offset + data.size
        if (required > capacity) resize(required)
        buffer!!.position(offset.toInt())
        buffer!!.put(data)
    }

    fun read(offset: Long, length: Int): ByteArray {
        ensureOpen()
        val buf = ByteArray(length)
        buffer!!.position(offset.toInt())
        buffer!!.get(buf)
        return buf
    }

    fun finalize(finalSize: Long) {
        closeMapping()
        RandomAccessFile(path, "rw").use { it.setLength(finalSize) }
        open(finalSize)
        buffer!!.force()
    }

    override fun close() {
        closeMapping()
    }

    // ── Internal ────────────────────────────────────────────────────────

    private fun open(size: Long) {
        file = RandomAccessFile(path, "rw")
        channel = file!!.channel
        buffer = channel!!.map(FileChannel.MapMode.READ_WRITE, 0, size)
        capacity = size
    }

    private fun ensureOpen() {
        if (buffer != null) return
        if (!Files.exists(Paths.get(path))) throw java.io.FileNotFoundException("Cache file not found: $path")
        val size = java.io.File(path).length()
        open(size)
    }

    private fun closeMapping() {
        buffer = null
        channel?.close(); channel = null
        file?.close(); file = null
    }

    private fun resize(required: Long) {
        var newSize = capacity
        while (newSize < required) newSize = minOf(newSize * 2, MAX_CACHE_SIZE)
        if (newSize > MAX_CACHE_SIZE) throw IllegalStateException("Cache exceeds max size")
        closeMapping()
        RandomAccessFile(path, "rw").use { it.setLength(newSize) }
        open(newSize)
    }
}
